File: elixir/poller.exs
# ─────────────────────────────────────────
# ALPR Poller — Elixir
# Camera: Prama PT-NC140D3-WNM(D2)
# ─────────────────────────────────────────
# Install: https://elixir-lang.org/install.html
#          mix local.hex && mix archive.install hex mix_install_preview
# Run:     elixir poller.exs

Mix.install([
  {:req, "~> 0.5"},
])

defmodule Poller do
  # ── Config (edit these)
  @rtsp_url    "rtsp://{username}:{password}@{your_ip}:{port}/Streaming/channels/101"
  @api_key     "your key, if any"
  @api_base    "your api url"
  @country     "IND"
  @interval_ms 2_000
  @cooldown_ms 8_000
  @max_retries 3

  @curr_frame "/tmp/alpr_curr.jpg"
  @prev_frame "/tmp/alpr_prev.jpg"

  # ── Capture one JPEG frame via ffmpeg
  defp capture_frame do
    if File.exists?(@curr_frame), do: File.copy!(@curr_frame, @prev_frame)

    case System.cmd("ffmpeg", [
      "-loglevel", "error",
      "-rtsp_transport", "tcp",
      "-i", @rtsp_url,
      "-frames:v", "1",
      "-q:v", "3",
      "-vf", "scale=640:480",
      @curr_frame, "-y"
    ], stderr_to_stdout: true) do
      {_, 0} ->
        current  = File.read!(@curr_frame)
        previous = if File.exists?(@prev_frame), do: File.read!(@prev_frame), else: nil
        {:ok, current, previous}

      {out, code} ->
        {:error, "ffmpeg exited #{code}: #{out}"}
    end
  end

  # ── Motion detection: byte diff score
  defp detect_motion(_current, nil), do: false
  defp detect_motion(current, previous) do
    if :crypto.hash(:md5, current) == :crypto.hash(:md5, previous) do
      false
    else
      len     = min(byte_size(current), byte_size(previous))
      indices = Enum.take_every(0..(len - 1), 10)
      diff    = Enum.reduce(indices, 0.0, fn i, acc ->
        acc + abs(:binary.at(current, i) - :binary.at(previous, i))
      end)
      score = diff / length(indices) / 255.0 * 100.0
      IO.puts("[motion] #{Float.round(score, 2)}%")
      score > 30
    end
  end

  # ── POST frame to ALPR API
  defp scan_plate do
    image = File.read!(@curr_frame)
    resp  = Req.post!(
      "#{@api_base}/plate/recognize",
      headers: [{"x-api-key", @api_key}],
      form_multipart: [
        image:   {image, filename: "frame.jpg", content_type: "image/jpeg"},
        country: @country,
      ]
    )
    if resp.status >= 300, do: raise("ALPR API #{resp.status}")
    resp.body
  end

  # ── Notify backend after confirmed plate
  defp notify_backend(result) do
    Req.post!(
      String.replace(@api_base, "/api/v1", "/api/entry"),
      headers: [{"x-api-key", @api_key}],
      json: %{
        plate:      result["plate"],
        confidence: result["confidence"],
        country:    @country,
        timestamp:  DateTime.utc_now() |> DateTime.to_iso8601(),
        source:     "elixir-poller-v1",
      }
    )
  end

  # ── Main loop
  def run do
    IO.puts("[alpr] elixir poller started")
    IO.puts("[camera] #{@rtsp_url}")
    loop(%{last_plate_ms: 0, errors: 0})
  end

  defp loop(state) do
    t0 = System.monotonic_time(:millisecond)

    state =
      cond do
        System.monotonic_time(:millisecond) - state.last_plate_ms < @cooldown_ms ->
          IO.puts("[cooldown] skipping")
          :timer.sleep(@interval_ms)
          state

        true ->
          state = case capture_frame() do
            {:error, reason} ->
              errors = state.errors + 1
              IO.puts("[error] ##{errors}: #{reason}")
              if errors >= @max_retries do
                IO.puts("[error] too many failures, waiting 30s")
                :timer.sleep(30_000)
                %{state | errors: 0}
              else
                %{state | errors: errors}
              end

            {:ok, current, previous} ->
              state = %{state | errors: 0}
              if not detect_motion(current, previous) do
                IO.puts("[motion] none")
                state
              else
                IO.puts("[motion] detected — scanning plate")
                try do
                  result = scan_plate()
                  IO.puts("[alpr] #{inspect(result)}")
                  plate = result["plate"]
                  if plate && plate != "" do
                    IO.puts("[plate] #{plate} (#{result["confidence"]}%)")
                    notify_backend(result)
                    IO.puts("[cooldown] #{@cooldown_ms / 1000}s")
                    %{state | last_plate_ms: System.monotonic_time(:millisecond)}
                  else
                    IO.puts("[plate] none found")
                    state
                  end
                rescue
                  e -> IO.puts("[error] #{Exception.message(e)}"); state
                end
              end
          end

          elapsed = System.monotonic_time(:millisecond) - t0
          sleep   = max(0, @interval_ms - elapsed)
          :timer.sleep(sleep)
          state
      end

    loop(state)
  end
end

Poller.run()

File: nodejs/poller.js
// ─────────────────────────────────────────
// ALPR Poller — Node.js  ★ OPTIMIZED
// Camera: Prama PT-NC140D3-WNM(D2)
// ─────────────────────────────────────────
// Single file, zero npm dependencies (requires Node.js >= 18).
// fetch and FormData are built-in from Node 18 — no npm install needed.
//
// Run: node --input-type=module poller.js
//  or: rename to poller.mjs then  node poller.mjs

import { execSync } from 'child_process';
import { existsSync, copyFileSync, readFileSync } from 'fs';
import { createHash } from 'crypto';

// ── Config (edit these)
const RTSP_URL         = 'rtsp://{username}:{password}@{your_ip}:{port}/Streaming/channels/101';
const API_KEY          = 'your key, if any';
const API_BASE         = 'your api url';
const COUNTRY          = 'IND';
const INTERVAL_MS      = 2000;
const COOLDOWN_MS      = 8000;
const MOTION_THRESHOLD = 30;
const MAX_RETRIES      = 3;

const CURR_FRAME = '/tmp/alpr_curr.jpg';
const PREV_FRAME = '/tmp/alpr_prev.jpg';

const sleep = ms => new Promise(r => setTimeout(r, ms));

// ── Capture one JPEG frame via ffmpeg
function captureFrame() {
  if (existsSync(CURR_FRAME)) copyFileSync(CURR_FRAME, PREV_FRAME);

  execSync(
    `ffmpeg -loglevel error -rtsp_transport tcp \
     -i "${RTSP_URL}" \
     -frames:v 1 -q:v 3 -vf scale=640:480 ${CURR_FRAME} -y`,
    { timeout: 6000 },
  );

  return {
    current:  readFileSync(CURR_FRAME),
    previous: existsSync(PREV_FRAME) ? readFileSync(PREV_FRAME) : null,
  };
}

// ── Motion detection: MD5 hash check + sampled byte diff
function detectMotion(current, previous) {
  if (!previous) return false;

  const h1 = createHash('md5').update(current).digest('hex');
  const h2 = createHash('md5').update(previous).digest('hex');
  if (h1 === h2) return false;

  const len = Math.min(current.length, previous.length);
  let diff = 0;
  for (let i = 0; i < len; i += 10) diff += Math.abs(current[i] - previous[i]);
  const score = (diff / (len / 10)) / 255 * 100;
  console.log(`[motion] ${score.toFixed(2)}%`);
  return score > MOTION_THRESHOLD;
}

// ── POST frame to ALPR API (uses built-in fetch + FormData from Node 18)
async function scanPlate() {
  const form = new FormData();
  form.append('image', new Blob([readFileSync(CURR_FRAME)], { type: 'image/jpeg' }), 'frame.jpg');
  form.append('country', COUNTRY);

  const res = await fetch(`${API_BASE}/plate/recognize`, {
    method:  'POST',
    headers: { 'x-api-key': API_KEY },
    body:    form,
  });
  if (!res.ok) throw new Error(`ALPR API ${res.status}`);
  return res.json();
}

// ── Notify backend after confirmed plate
async function notifyBackend(result) {
  await fetch(`${API_BASE}/entry`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json', 'x-api-key': API_KEY },
    body:    JSON.stringify({
      plate:      result.plate,
      confidence: result.confidence,
      country:    COUNTRY,
      timestamp:  new Date().toISOString(),
      source:     'node-poller-v3',
    }),
  });
}

// ── Main loop
async function main() {
  console.log('[alpr] node.js poller started');
  console.log(`[camera] ${RTSP_URL}`);

  let lastPlateTime     = 0;
  let consecutiveErrors = 0;

  while (true) {
    const t0 = Date.now();
    try {
      if (Date.now() - lastPlateTime < COOLDOWN_MS) {
        console.log('[cooldown] skipping');
        await sleep(INTERVAL_MS);
        continue;
      }

      const { current, previous } = captureFrame();
      consecutiveErrors = 0;

      if (!detectMotion(current, previous)) {
        console.log('[motion] none');
      } else {
        console.log('[motion] detected — scanning plate');
        const result = await scanPlate();
        console.log('[alpr]', JSON.stringify(result));

        if (result?.plate) {
          console.log(`[plate] ${result.plate} (${result.confidence}%)`);
          await notifyBackend(result);
          lastPlateTime = Date.now();
          console.log(`[cooldown] ${COOLDOWN_MS / 1000}s`);
        } else {
          console.log('[plate] none found');
        }
      }
    } catch (err) {
      consecutiveErrors++;
      console.error(`[error] #${consecutiveErrors}: ${err.message}`);
      if (consecutiveErrors >= MAX_RETRIES) {
        console.error('[error] too many failures, waiting 30s');
        await sleep(30_000);
        consecutiveErrors = 0;
      }
    }

    const elapsed = Date.now() - t0;
    if (elapsed < INTERVAL_MS) await sleep(INTERVAL_MS - elapsed);
  }
}

main();
